File: sdk/ios/Sources/Afftok/ClickTracker.swift
import Foundation
import CommonCrypto

/// Click tracking implementation for AffTok SDK
public class ClickTracker {
    
    private let options: AfftokOptions
    private let queue: OfflineQueue
    private let session: URLSession
    
    public init(options: AfftokOptions, queue: OfflineQueue) {
        self.options = options
        self.queue = queue
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = AfftokConfig.connectionTimeout
        config.timeoutIntervalForResource = AfftokConfig.readTimeout
        self.session = URLSession(configuration: config)
    }
    
    /// Track a click event
    public func trackClick(params: ClickParams) async -> AfftokResponse {
        let payload = buildClickPayload(params: params)
        
        do {
            let response = try await sendRequest(endpoint: AfftokConfig.clickEndpoint, payload: payload)
            if response.success {
                log("Click tracked successfully: \(params.offerId)")
            } else {
                queue.enqueue(type: "click", payload: payload)
                log("Click queued for retry: \(params.offerId)")
            }
            return response
        } catch {
            queue.enqueue(type: "click", payload: payload)
            log("Click queued (offline): \(params.offerId), error: \(error.localizedDescription)")
            return AfftokResponse(
                success: false,
                message: "Click queued for offline retry",
                error: error.localizedDescription
            )
        }
    }
    
    /// Track click with signed link
    public func trackSignedClick(signedLink: String, params: ClickParams) async -> AfftokResponse {
        var payload = buildClickPayload(params: params)
        payload["signed_link"] = signedLink
        payload["link_validated"] = "true"
        
        do {
            let response = try await sendRequest(endpoint: AfftokConfig.clickEndpoint, payload: payload)
            if !response.success {
                queue.enqueue(type: "click", payload: payload)
            }
            return response
        } catch {
            queue.enqueue(type: "click", payload: payload)
            return AfftokResponse(
                success: false,
                message: "Signed click queued for retry",
                error: error.localizedDescription
            )
        }
    }
    
    /// Build click payload
    private func buildClickPayload(params: ClickParams) -> [String: Any] {
        let timestamp = Int64(Date().timeIntervalSince1970 * 1000)
        let nonce = generateNonce()
        let deviceInfo = DeviceInfo.getDeviceInfo()
        
        var payload: [String: Any] = [
            "api_key": options.apiKey,
            "advertiser_id": options.advertiserId,
            "offer_id": params.offerId,
            "timestamp": timestamp,
            "nonce": nonce,
            "device_info": deviceInfo
        ]
        
        if let userId = options.userId {
            payload["user_id"] = userId
        }
        if let trackingCode = params.trackingCode {
            payload["tracking_code"] = trackingCode
        }
        if let subId1 = params.subId1 {
            payload["sub_id_1"] = subId1
        }
        if let subId2 = params.subId2 {
            payload["sub_id_2"] = subId2
        }
        if let subId3 = params.subId3 {
            payload["sub_id_3"] = subId3
        }
        if let customParams = params.customParams {
            payload["custom_params"] = customParams
        }
        
        // Add signature
        payload["signature"] = generateSignature(timestamp: timestamp, nonce: nonce)
        
        return payload
    }
    
    /// Send HTTP request
    private func sendRequest(endpoint: String, payload: [String: Any]) async throws -> AfftokResponse {
        guard let url = URL(string: "\(options.baseUrl)\(endpoint)") else {
            throw URLError(.badURL)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(options.apiKey, forHTTPHeaderField: "X-API-Key")
        request.setValue(AfftokConfig.sdkVersion, forHTTPHeaderField: "X-SDK-Version")
        request.setValue(AfftokConfig.sdkPlatform, forHTTPHeaderField: "X-SDK-Platform")
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        
        if httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 {
            let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            return AfftokResponse(
                success: true,
                message: json?["message"] as? String,
                data: json
            )
        } else {
            // Try fallback endpoint
            if endpoint == AfftokConfig.clickEndpoint {
                return try await sendFallbackRequest(endpoint: AfftokConfig.fallbackClickEndpoint, payload: payload)
            }
            return AfftokResponse(
                success: false,
                error: "HTTP \(httpResponse.statusCode)"
            )
        }
    }
    
    /// Send request to fallback endpoint
    private func sendFallbackRequest(endpoint: String, payload: [String: Any]) async throws -> AfftokResponse {
        guard let url = URL(string: "\(options.baseUrl)\(endpoint)") else {
            throw URLError(.badURL)
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(options.apiKey, forHTTPHeaderField: "X-API-Key")
        request.httpBody = try JSONSerialization.data(withJSONObject: payload)
        
        let (_, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        
        if httpResponse.statusCode >= 200 && httpResponse.statusCode < 300 {
            return AfftokResponse(success: true, message: "Click tracked via fallback")
        } else {
            return AfftokResponse(success: false, error: "Fallback failed: \(httpResponse.statusCode)")
        }
    }
    
    /// Generate HMAC-SHA256 signature
    private func generateSignature(timestamp: Int64, nonce: String) -> String {
        let dataToSign = "\(options.apiKey)|\(options.advertiserId)|\(timestamp)|\(nonce)"
        
        guard let keyData = options.apiKey.data(using: .utf8),
              let messageData = dataToSign.data(using: .utf8) else {
            return ""
        }
        
        var hmac = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        
        keyData.withUnsafeBytes { keyBytes in
            messageData.withUnsafeBytes { messageBytes in
                CCHmac(CCHmacAlgorithm(kCCHmacAlgSHA256),
                       keyBytes.baseAddress, keyData.count,
                       messageBytes.baseAddress, messageData.count,
                       &hmac)
            }
        }
        
        return hmac.map { String(format: "%02x", $0) }.joined()
    }
    
    /// Generate random nonce
    private func generateNonce() -> String {
        let chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
        return String((0..<32).map { _ in chars.randomElement()! })
    }
    
    private func log(_ message: String) {
        if options.debug {
            print("[AffTok Click] \(message)")
        }
    }
}

File: sdk/android/afftok-sdk/src/main/java/com/afftok/ClickTracker.kt
package com.afftok

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Click tracking implementation for AffTok SDK
 */
class ClickTracker(
    private val options: AfftokOptions,
    private val fingerprintProvider: FingerprintProvider,
    private val queue: OfflineQueue
) {
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(Config.CONNECTION_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        .readTimeout(Config.READ_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        .writeTimeout(Config.WRITE_TIMEOUT_MS, TimeUnit.MILLISECONDS)
        .build()
    
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()
    
    /**
     * Track a click event
     */
    suspend fun trackClick(params: ClickParams): AfftokResponse {
        val payload = buildClickPayload(params)
        
        return withContext(Dispatchers.IO) {
            try {
                val response = sendRequest(Config.CLICK_ENDPOINT, payload)
                if (response.success) {
                    log("Click tracked successfully: ${params.offerId}")
                } else {
                    // Queue for retry
                    queue.enqueue("click", payload)
                    log("Click queued for retry: ${params.offerId}")
                }
                response
            } catch (e: Exception) {
                // Queue for offline retry
                queue.enqueue("click", payload)
                log("Click queued (offline): ${params.offerId}, error: ${e.message}")
                AfftokResponse(
                    success = false,
                    error = e.message,
                    message = "Click queued for offline retry"
                )
            }
        }
    }
    
    /**
     * Track click with signed link
     */
    suspend fun trackSignedClick(signedLink: String, params: ClickParams): AfftokResponse {
        val payload = buildClickPayload(params).toMutableMap()
        payload["signed_link"] = signedLink
        payload["link_validated"] = true
        
        return withContext(Dispatchers.IO) {
            try {
                val response = sendRequest(Config.CLICK_ENDPOINT, payload)
                if (!response.success) {
                    queue.enqueue("click", payload)
                }
                response
            } catch (e: Exception) {
                queue.enqueue("click", payload)
                AfftokResponse(
                    success = false,
                    error = e.message,
                    message = "Signed click queued for retry"
                )
            }
        }
    }
    
    /**
     * Build click payload
     */
    private fun buildClickPayload(params: ClickParams): Map<String, Any> {
        val timestamp = System.currentTimeMillis()
        val nonce = generateNonce()
        val deviceInfo = fingerprintProvider.getDeviceInfo()
        
        val payload = mutableMapOf<String, Any>(
            "api_key" to options.apiKey,
            "advertiser_id" to options.advertiserId,
            "offer_id" to params.offerId,
            "timestamp" to timestamp,
            "nonce" to nonce,
            "device_info" to deviceInfo
        )
        
        options.userId?.let { payload["user_id"] = it }
        params.trackingCode?.let { payload["tracking_code"] = it }
        params.subId1?.let { payload["sub_id_1"] = it }
        params.subId2?.let { payload["sub_id_2"] = it }
        params.subId3?.let { payload["sub_id_3"] = it }
        params.customParams?.let { payload["custom_params"] = it }
        
        // Add signature
        payload["signature"] = generateSignature(payload, timestamp, nonce)
        
        return payload
    }
    
    /**
     * Send HTTP request
     */
    private fun sendRequest(endpoint: String, payload: Map<String, Any>): AfftokResponse {
        val url = "${options.baseUrl}$endpoint"
        val json = JSONObject(payload).toString()
        val body = json.toRequestBody(jsonMediaType)
        
        val request = Request.Builder()
            .url(url)
            .post(body)
            .addHeader("Content-Type", "application/json")
            .addHeader("X-API-Key", options.apiKey)
            .addHeader("X-SDK-Version", Config.SDK_VERSION)
            .addHeader("X-SDK-Platform", Config.SDK_PLATFORM)
            .build()
        
        val response = client.newCall(request).execute()
        val responseBody = response.body?.string()
        
        return if (response.isSuccessful) {
            val jsonResponse = responseBody?.let { JSONObject(it) }
            AfftokResponse(
                success = true,
                message = jsonResponse?.optString("message"),
                data = jsonResponse?.let { parseJsonToMap(it) }
            )
        } else {
            // Try fallback endpoint
            if (endpoint == Config.CLICK_ENDPOINT) {
                return sendFallbackRequest(Config.FALLBACK_CLICK_ENDPOINT, payload)
            }
            AfftokResponse(
                success = false,
                error = "HTTP ${response.code}: $responseBody"
            )
        }
    }
    
    /**
     * Send request to fallback endpoint
     */
    private fun sendFallbackRequest(endpoint: String, payload: Map<String, Any>): AfftokResponse {
        return try {
            val url = "${options.baseUrl}$endpoint"
            val json = JSONObject(payload).toString()
            val body = json.toRequestBody(jsonMediaType)
            
            val request = Request.Builder()
                .url(url)
                .post(body)
                .addHeader("Content-Type", "application/json")
                .addHeader("X-API-Key", options.apiKey)
                .build()
            
            val response = client.newCall(request).execute()
            
            if (response.isSuccessful) {
                AfftokResponse(success = true, message = "Click tracked via fallback")
            } else {
                AfftokResponse(success = false, error = "Fallback failed: ${response.code}")
            }
        } catch (e: Exception) {
            AfftokResponse(success = false, error = "Fallback error: ${e.message}")
        }
    }
    
    /**
     * Generate HMAC-SHA256 signature
     */
    private fun generateSignature(payload: Map<String, Any>, timestamp: Long, nonce: String): String {
        val dataToSign = "${options.apiKey}|${options.advertiserId}|$timestamp|$nonce"
        
        return try {
            val mac = Mac.getInstance("HmacSHA256")
            val secretKey = SecretKeySpec(options.apiKey.toByteArray(), "HmacSHA256")
            mac.init(secretKey)
            val hmacBytes = mac.doFinal(dataToSign.toByteArray())
            hmacBytes.joinToString("") { "%02x".format(it) }
        } catch (e: Exception) {
            log("Signature generation failed: ${e.message}")
            ""
        }
    }
    
    /**
     * Generate random nonce
     */
    private fun generateNonce(): String {
        val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
        return (1..32).map { chars.random() }.joinToString("")
    }
    
    /**
     * Parse JSON to Map
     */
    private fun parseJsonToMap(json: JSONObject): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        json.keys().forEach { key ->
            map[key] = json.get(key)
        }
        return map
    }
    
    private fun log(message: String) {
        if (options.debug) {
            println("[AffTok Click] $message")
        }
    }
}

